pkg/services/store/entity/sqlstash/sql_storage_server.go

```diff
@@ -1486,8 +1582,345 @@ func watchMatches(r *entity.EntityWatchRequest, result *entity.Entity) bool {
+// are sure we are interested in. Notifications include a payload which aids discriminating which are relevant to a
+// watcher.
+func (s *sqlEntityServer) pgWatch(ctx context.Context, r cachedEntityWatchRequest, w entity.EntityStore_WatchServer) error {
+	// create a channel for the broker to send us the notifications we're interested in and register it
+	c := make(chan *pq.Notification, s.pgWatcherChannelBufferLen)
+	watcherKey, err := s.pgAddNotifyChan(r.namespace, c)
+	if err != nil {
+		return err
+	}
+
+	// cleanup
+	defer func() {
+		if err := s.pgRemoveNotifyChan(r.namespace, watcherKey); err != nil {
+			s.log.Error("remove notify channel", "error", err)
+		}
+		close(c)
+	}()
+
+	// main notifications loop
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			// NB: if you call `break` within a `select` statement you will break the select, so we need a label to
+			// explicitly refer to the loop
+			break loop
+
+		case n, ok := <-c:
+			if !ok {
+				// we were terminated by pgTerminateWatchers
+				return ErrPgWatchTerminated
+			}
+
+			// decode the notification payload
+			var ne notifiedEntity
+			if err := json.NewDecoder(strings.NewReader(n.Extra)).Decode(&ne); err != nil {
+				s.log.Error("decode postgres notification from channel %q with JSON payload %q: %w", r.namespace, n.Extra, err)
+				continue
+			}
+
+			// continue if we don't care about this notification
+			partialEntity, err := ne.toEntity(r.namespace)
+			if err != nil {
+				s.log.Error("convert notification with payload %q to entity: %w", n.Extra, err)
+				continue
+			}
+			if !watchMatches(r, partialEntity) {
+				continue
+			}
+
+			// retrieve the full row from the database. We are using Query instead of Get method to reuse rowToEntity
+			// for max consistency, but we are sure we cannot get more than one row since the `guid` field is the
+			// primary key of the table
+			rows, err := s.sess.Query(ctx, "SELECT * FROM entity_history WHERE guid = ?", ne.GUID)
+			if err != nil {
+				return fmt.Errorf("get from entity history by GUID %q: %w", ne.GUID, err)
+			}
+			defer rows.Close()
+
+			// advance the rows iterator to retrieve the row
+			if !rows.Next() {
+				return fmt.Errorf("entity with GUID %q not found in history", ne.GUID)
+			}
+
+			// convert the row to entity
+			result, err := s.rowToEntity(ctx, rows, r.WithBody, r.WithStatus)
+			if err != nil {
+				return fmt.Errorf("convert SQL row to entity with GUID %q: %w", ne.GUID, err)
+			}
+
+			// send the results to the client
+			s.log.Debug("sending watch result", "guid", result.Guid, "action", result.Action)
+			err = w.Send(&entity.EntityWatchResponse{
+				Timestamp: time.Now().UnixMilli(),
+				Entity:    result,
+			})
+			if err != nil {
+				return fmt.Errorf("sending entity with GUID %q to client: %w", ne.GUID, err)
+			}
+
+			// update r.Since value so we don't send earlier results again
+			r.Since = result.ResourceVersion
+		}
+	}
+
+	return nil
+}
+
```

Comment:
there seems to be a lot of duplication between this code and the existing poll-based setup, is there a reason not to just update line ~147 to choose between the standard poller and the pg watch implementation, but keep the same broadcast code?

Reply:
Actually, I think I can work around it to keep track of the postgres channels listened to separately and unlisten and terminate separately, and that would make the code more compatible with Broadcaster. I'll try to do that, but I'd like to wrap a bit first for the hackathon video as well. But I also think we may need the unbounded buffer channel implementation for the reasons above, wdyt?